src/Datadog.Trace.ClrProfiler.Native/clr_helpers.cpp

```diff
@@ -999,4 +1023,553 @@ bool ReturnTypeIsValueTypeOrGeneric(
+        }
+      }
+      tokenName += "]"_W;
+      break;
+    }
+    case ELEMENT_TYPE_MVAR: {
+      pbCur++;
+      ULONG num = 0;
+      pbCur += CorSigUncompressData(pbCur, &num);
+      tokenName = "!!"_W + ToWSTRING(std::to_string(num));
+      break;
+    }
+    case ELEMENT_TYPE_VAR: {
+      pbCur++;
+      ULONG num = 0;
+      pbCur += CorSigUncompressData(pbCur, &num);
+      tokenName = "!"_W + ToWSTRING(std::to_string(num));
+      break;
+    }
+    default:
+      break;
+  }
+
+  if (ref_flag) {
+    tokenName += "&"_W;
+  }
+  return tokenName;
+}
+
+WSTRING FunctionMethodArgument::GetTypeTokName(ComPtr<IMetaDataImport2>& pImport) const {
+  PCCOR_SIGNATURE pbCur = &pbBase[offset];
+  return GetSigTypeTokName(pbCur, pImport);
+}
+
+ULONG FunctionMethodArgument::GetSignature(PCCOR_SIGNATURE& data) const {
+  data = &pbBase[offset];
+  return length;
+}
+
+// FunctionMethodSignature
+bool ParseByte(PCCOR_SIGNATURE& pbCur, PCCOR_SIGNATURE pbEnd, unsigned char* pbOut) {
+  if (pbCur < pbEnd) {
+    *pbOut = *pbCur;
+    pbCur++;
+    return true;
+  }
+
+  return false;
+}
+
+bool ParseNumber(PCCOR_SIGNATURE& pbCur, PCCOR_SIGNATURE pbEnd, unsigned* pOut) {
+  // parse the variable length number format (0-4 bytes)
+
+  unsigned char b1 = 0, b2 = 0, b3 = 0, b4 = 0;
+
+  // at least one byte in the encoding, read that
+
+  if (!ParseByte(pbCur, pbEnd, &b1)) return false;
+
+  if (b1 == 0xff) {
+    // special encoding of 'NULL'
+    // not sure what this means as a number, don't expect to see it except for
+    // string lengths which we don't encounter anyway so calling it an error
+    return false;
+  }
+
+  // early out on 1 byte encoding
+  if ((b1 & 0x80) == 0) {
+    *pOut = (int)b1;
+    return true;
+  }
+
+  // now at least 2 bytes in the encoding, read 2nd byte
+  if (!ParseByte(pbCur, pbEnd, &b2)) return false;
+
+  // early out on 2 byte encoding
+  if ((b1 & 0x40) == 0) {
+    *pOut = (((b1 & 0x3f) << 8) | b2);
+    return true;
+  }
+
+  // must be a 4 byte encoding
+  if ((b1 & 0x20) != 0) {
+    // 4 byte encoding has this bit clear -- error if not
+    return false;
+  }
+
+  if (!ParseByte(pbCur, pbEnd, &b3)) return false;
+
+  if (!ParseByte(pbCur, pbEnd, &b4)) return false;
+
+  *pOut = ((b1 & 0x1f) << 24) | (b2 << 16) | (b3 << 8) | b4;
+  return true;
+}
+
+bool ParseTypeDefOrRefEncoded(PCCOR_SIGNATURE& pbCur, PCCOR_SIGNATURE pbEnd,
+                              unsigned char* pIndexTypeOut,
+                              unsigned* pIndexOut) {
+  // parse an encoded typedef or typeref
+  unsigned encoded = 0;
+
+  if (!ParseNumber(pbCur, pbEnd, &encoded)) return false;
+
+  *pIndexTypeOut = (unsigned char)(encoded & 0x3);
+  *pIndexOut = (encoded >> 2);
+  return true;
+}
+
+/*  we don't support
+    PTR CustomMod* VOID
+    PTR CustomMod* Type
+    FNPTR MethodDefSig
+    FNPTR MethodRefSig
+    ARRAY Type ArrayShape
+    CustomMod*
+ */
+bool ParseType(PCCOR_SIGNATURE& pbCur, PCCOR_SIGNATURE pbEnd) {
```

Comment:
I thought this looked familiar, and I realized I also did a version of this in `sig_helpers.cpp`. After this PR let's consolidate the usage of all `ParseType` calls to your version because I'm guessing it's more thorough than the previous version that I wrote

Reply:
Sure, I didn't want to handle any modification to the current implementation to avoid modifying behavior or something else. What I did in some cases is add members to an struct or class or use a new type. Later we can think of merging things.